### PR TITLE
Adjust Claudio's raw SHA2 formats to deal with GPU mask mode benchmark.

### PR DESCRIPTION
**Pull request by claudioandre**
* A PR just to think about it.

### CODE REVIEW

**Comment by claudioandre** [comment]
1. below, something that seems to be an improvement. Please, review and analyse.

```
diff --git a/src/common-opencl.c b/src/common-opencl.c
index ce5e361..852f506 100644
--- a/src/common-opencl.c
+++ b/src/common-opencl.c
@@ -1448,9 +1459,9 @@ void opencl_find_best_lws(size_t group_size_limit, int sequential_id,
 			break;
 		if (options.verbosity > 3)
 			fprintf(stderr, " %s%s\n", ns2string(sumEndTime - sumStartTime),
-			    ((double)(sumEndTime - sumStartTime) / kernelExecTimeNs < 0.997)
+			    ((double)(sumEndTime - sumStartTime) / kernelExecTimeNs < 0.999)
 			        ? "+" : "");
-		if ((double)(sumEndTime - sumStartTime) / kernelExecTimeNs < 0.997) {
+		if ((double)(sumEndTime - sumStartTime) / kernelExecTimeNs < 0.999) {
 			kernelExecTimeNs = sumEndTime - sumStartTime;
 			optimal_work_group = my_work_group;
 		} else {
```

**Comment by magnumripper** [comment]
Doesn't this make a mere 0.1% improvement get picked as "better"? I thought 0.3% was quite low already: The general idea (I think) is that a lower LWS and/or GWS is always preferable when they perform the same. But if you've seen it end up (after all passes) with a significantly better pair of figures, it might be worthwhile.

File: src/opencl_rawsha256_fmt_plug.c [hunk 1]
```diff
@@ -304,18 +309,21 @@ static void reset(struct db_main *db)
...
 
-		if (options.flags & FLG_MASK_CHK) {
+		if ((options.flags & FLG_MASK_CHK) && !(options.flags & FLG_TEST_CHK)) {
 			fprintf(stdout,
 			        "Using Mask Mode with internal candidate generation%s",
 			        flag ? "" : "\n");
 
 			if (flag) {
-				self->params.max_keys_per_crypt /= 256;
+				self->params.max_keys_per_crypt /= 32;
```
Comment (magnumripper) [inline src/opencl_rawsha256_fmt_plug.c]: If any division is needed here, I think it indicates some fundamental flaw in the shared code. We should try to fix that instead.

Comment (claudioandre) [inline src/opencl_rawsha256_fmt_plug.c]: > I think it indicates some fundamental flaw

Yes, there is. Without --test is does execute a "regular run". I haven't seen why.

**Comment by claudioandre** [comment]
> Doesn't this make a mere 0.1% improvement get picked as "better"? 

Below, why I did the suggestion. This happens only on super's -dev=7 (only for 32->64). 
```
[...]
Calculating best LWS for GWS=4096
Testing LWS=32 GWS=4096 ... 187.320ms+
Testing LWS=64 GWS=4096 ... 187.196ms
[...]
```

So, 32 was picked, but 64 is much better on a real run. 
* I saw a benefit and no drawbacks.

File: src/opencl_rawsha256_fmt_plug.c (hunk 1 above)
Comment (magnumripper) [inline src/opencl_rawsha256_fmt_plug.c]: It works correctly with eg. nt-opencl (which uses our shared auto-tune). Check for clues in reset().

**Comment by magnumripper** [comment]
I'd rather change default of AutotuneLWS in john.conf to 0 (== start with a LWS=NULL run). That usually ends up better. Solar thought it was slow (relatively speaking) but I normally get better results with it.

**Comment by magnumripper** [comment]
AutotuneLWS = 0 doesn't seem to make any difference in this case though.

**Comment by magnumripper** [comment]
OK, after all this we still have your problem
```
$ ../run/john -test -form:raw-sha256-opencl -v:4 -dev=7 -mask
Device 7: GeForce GTX TITAN
Benchmarking: Raw-SHA256-opencl [SHA256 OpenCL]... Options used: -I ../run/kernels -cl-mad-enable -DSM_MAJOR=3 -DSM_MINOR=5 -cl-nv-verbose -D__GPU__ -DDEVICE_INFO=131090 -DDEV_VER_MAJOR=352 -DDEV_VER_MINOR=21 -D_OPENCL_COMPILER  $JOHN/kernels/sha256_kernel.cl
Internal mask, multiplier: 10000 (target: 8547)
Calculating best GWS for LWS=32; max. 500ms single kernel invocation.
gws:      1024	 47595Kc/s    47595865 rounds/s 215.144ms per crypt_all()!
gws:      2048	103911Kc/s   103911150 rounds/s 197.091ms per crypt_all()!
gws:      4096	197074Kc/s   197074248 rounds/s 207.840ms per crypt_all()+
gws:      8192	182377Kc/s   182377347 rounds/s 449.178ms per crypt_all()
Calculating best LWS for GWS=4096
Testing LWS=32 GWS=4096 ... 207.894ms+
Testing LWS=64 GWS=4096 ... 207.638ms
Testing LWS=128 GWS=4096 ... 212.061ms
Testing LWS=256 GWS=4096 ... 230.923ms
Testing LWS=512 GWS=4096 ... 237.994ms
Calculating best GWS for LWS=32; max. 500ms single kernel invocation.
gws:       448	 22677Kc/s    22677264 rounds/s 197.554ms per crypt_all()!
gws:       896	 47726Kc/s    47726708 rounds/s 187.735ms per crypt_all()!
gws:      1792	 97984Kc/s    97984288 rounds/s 182.886ms per crypt_all()!
gws:      3584	189617Kc/s   189617686 rounds/s 189.011ms per crypt_all()+
gws:      7168	316902Kc/s   316902594 rounds/s 226.189ms per crypt_all()+
gws:     14336	316931Kc/s   316931491 rounds/s 452.337ms per crypt_all()
Local worksize (LWS) 32, global worksize (GWS) 7168
Using Mask Mode with internal candidate generation, global worksize(GWS) set to 28
DONE, GPU util:99%
Raw:	1419K c/s real, 1419K c/s virtual

$ LWS=64 ../run/john -test -form:raw-sha256-opencl -v:4 -dev=7 -mask
Device 7: GeForce GTX TITAN
Benchmarking: Raw-SHA256-opencl [SHA256 OpenCL]... Options used: -I ../run/kernels -cl-mad-enable -DSM_MAJOR=3 -DSM_MINOR=5 -cl-nv-verbose -D__GPU__ -DDEVICE_INFO=131090 -DDEV_VER_MAJOR=352 -DDEV_VER_MINOR=21 -D_OPENCL_COMPILER  $JOHN/kernels/sha256_kernel.cl
Internal mask, multiplier: 10000 (target: 8547)
Calculating best GWS for LWS=64; max. 500ms single kernel invocation.
gws:       896	 41604Kc/s    41604810 rounds/s 215.359ms per crypt_all()!
gws:      1792	 94219Kc/s    94219197 rounds/s 190.194ms per crypt_all()!
gws:      3584	189499Kc/s   189499783 rounds/s 189.129ms per crypt_all()!
gws:      7168	316691Kc/s   316691657 rounds/s 226.340ms per crypt_all()+
gws:     14336	471776Kc/s   471776249 rounds/s 303.872ms per crypt_all()+
Local worksize (LWS) 64, global worksize (GWS) 14336
Using Mask Mode with internal candidate generation, global worksize(GWS) set to 56
DONE, GPU util:99%
Raw:	2975K c/s real, 2975K c/s virtual
```

**Comment by magnumripper** [comment]
...but using LWS=NULL is a solution now:
```
$ LWS=0 ../run/john -test -form:raw-sha256-opencl -v:4 -dev=7 -mask
Device 7: GeForce GTX TITAN
Benchmarking: Raw-SHA256-opencl [SHA256 OpenCL]... Options used: -I ../run/kernels -cl-mad-enable -DSM_MAJOR=3 -DSM_MINOR=5 -cl-nv-verbose -D__GPU__ -DDEVICE_INFO=131090 -DDEV_VER_MAJOR=352 -DDEV_VER_MINOR=21 -D_OPENCL_COMPILER  $JOHN/kernels/sha256_kernel.cl
Internal mask, multiplier: 10000 (target: 8547)
Calculating best GWS for LWS=0; max. 500ms single kernel invocation.
gws:      1024	 48905Kc/s    48905674 rounds/s 209.382ms per crypt_all()!
gws:      2048	104543Kc/s   104543151 rounds/s 195.899ms per crypt_all()!
gws:      4096	193306Kc/s   193306198 rounds/s 211.891ms per crypt_all()+
gws:      8192	344483Kc/s   344483381 rounds/s 237.805ms per crypt_all()+
gws:     16384	440478Kc/s   440478531 rounds/s 371.959ms per crypt_all()+
Calculating best LWS for GWS=16384
Testing LWS=32 GWS=16384 ... 674.567ms+
Testing LWS=64 GWS=16384 ... 537.874ms+
Testing LWS=128 GWS=16384 ... 371.774ms+
Testing LWS=256 GWS=16384 ... 376.253ms
Testing LWS=512 GWS=16384 ... 660.795ms
Calculating best GWS for LWS=128; max. 500ms single kernel invocation.
gws:      1792	 97989Kc/s    97989037 rounds/s 182.877ms per crypt_all()!
gws:      3584	189486Kc/s   189486221 rounds/s 189.143ms per crypt_all()+
gws:      7168	316924Kc/s   316924788 rounds/s 226.173ms per crypt_all()+
gws:     14336	471615Kc/s   471615038 rounds/s 303.976ms per crypt_all()+
Local worksize (LWS) 128, global worksize (GWS) 14336
Using Mask Mode with internal candidate generation, global worksize(GWS) set to 56
DONE, GPU util:99%
Raw:	3049K c/s real, 3049K c/s virtual
```

**Comment by claudioandre** [comment]
It only happens in some situations, a good user can handle it.

**Comment by claudioandre** [comment]
The PR is working now.
* should I remove the opencl_get_tune_time_limit() I created?

**Comment by magnumripper** [comment]
Please do, my function seems to do the job fine.

**Comment by magnumripper** [comment]
...and I really think bumping duration is the wrong solution. Decreasing int_cands is much nicer.

**Comment by claudioandre** [comment]
Done, CI in progress.

**Pull request merged**